### PR TITLE
feat: add backend safeguard against deleting topic that has posts

### PR DESCRIPTION
**Pull request by zxt-tzx**
## Problem

Following https://github.com/opengovsg/askgovsg/pull/1350, I was somewhat uncomfortable with the fact that deleting a topic that will result in orphan posts (with no parent topic) is only safeguarded on the frontend. 

## Solution

Have modified the backend slightly to add an additional layer of safeguard

## Doubts

I'm not very familiar with Sequelize, so am not very sure if that part of `post.service.ts` is correct. Would appreciate if reviewer could take a closer look at that.

### CODE REVIEW

**Review by LoneRifle — CHANGES_REQUESTED** [review]
- a better name for the func would be `getPostsByTopic`
- could you provide a test that exercises the case when `hasPosts` on topic deletion?

**Comment by zxt-tzx** [comment]
done

File: server/src/modules/topics/topics.controller.ts [hunk 1]
```diff
@@ -169,6 +174,16 @@ export class TopicsController {
         .json({ message: 'You do not have permission to delete this topic' })
     }
 
+    // Check if topic has posts
+    const posts = await this.postService.getPostsByTopic(topicId)
+    const hasPosts = posts.totalItems !== 0
+
+    if (hasPosts) {
+      return res
+        .status(StatusCodes.FORBIDDEN)
```
Comment (LoneRifle) [inline server/src/modules/topics/topics.controller.ts]:
```suggestion
        .status(StatusCodes.BAD_REQUEST)
```
403 (and the related 401) deals with permission-related issues, not so much the correctness of a request

**Review by LoneRifle — CHANGES_REQUESTED** [review]
Just one more. FIx this up and adjust the tests as needed

File: server/src/modules/topics/topics.controller.ts (hunk 1 above)
Comment (zxt-tzx) [reply server/src/modules/topics/topics.controller.ts]: OK

Comment (zxt-tzx) [reply server/src/modules/topics/topics.controller.ts]: was debating which to use and tbh neither seemed like a perfect fit!
image
image

Comment (zxt-tzx) [reply server/src/modules/topics/topics.controller.ts]: done

**Review by LoneRifle — APPROVED** [review]

**Pull request merged**